utils/src/priority_queue.rs

```diff
@@ -0,0 +1,147 @@
+use std::cmp::Ordering;
+use std::collections::{BTreeSet, HashMap, HashSet};
+use std::hash::Hash;
+
+#[derive(Eq, PartialEq, Clone)]
+struct Entry<I: Ord + Hash + Clone, V: Ord + Clone> {
+    item: I,
+    value: V,
+}
+
+impl<I: Ord + Hash + Clone, V: Ord + Clone> Ord for Entry<I, V> {
+    fn cmp(&self, other: &Self) -> Ordering {
+        match self.value.cmp(&other.value) {
+            Ordering::Equal => self.item.cmp(&other.item),
+            other => other,
+        }
+    }
+}
+
+impl<I: Ord + Hash + Clone, V: Ord + Clone> PartialOrd for Entry<I, V> {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+/// A generic priority queue that enforces item uniqueness and optimizes for
+/// fast priority-ordered iteration rather than memory usage.
+pub struct PriorityQueue<I: Ord + Hash + Clone, V: Ord + Clone> {
+    entries: BTreeSet<Entry<I, V>>,
+    keys: HashMap<I, V>,
+}
+
+impl<I: Ord + Hash + Clone, V: Ord + Clone> PriorityQueue<I, V> {
+    /// Create a new priority queue.
+    #[allow(clippy::new_without_default)]
+    pub fn new() -> Self {
+        Self {
+            entries: BTreeSet::new(),
+            keys: HashMap::new(),
+        }
+    }
+
+    /// Insert an item with a value, overwriting the previous value if it exists.
+    pub fn put(&mut self, item: I, value: V) {
```

Comment:
The need to `clone` both `item` and `value` multiple times during `put` is suboptimal. We should make this better.